src/token/src/ledger/history.rs

```diff
@@ -0,0 +1,329 @@
+use crate::types::TxRecord;
+use candid::types::{Compound, Serializer, Type};
+use candid::Principal;
+use ic_cdk::export::candid::{CandidType, Deserialize, Nat};
+use ic_certified_map::{HashTree, RbTree};
+use serde::de::{SeqAccess, Visitor};
+use serde::Deserializer;
+use std::fmt;
+use std::fmt::Formatter;
+
+#[cfg(target_arch = "wasm32")]
+use ic_certified_map::AsHashTree;
+
+const MAX_TREE_SIZE: usize = 100_000;
+
+#[derive(Default, Clone)]
+pub struct History {
+    tree: RbTree<Vec<u8>, Vec<u8>>,
+    tree_size: usize,
+}
+
+impl History {
+    pub fn insert(&mut self, tx: TxRecord) {
+        self.tree
+            .insert(get_key_bytes(&tx.index), get_tx_bytes(&tx));
+        self.tree_size += 1;
+        if self.tree_size > MAX_TREE_SIZE {
+            self.remove_oldest_tx();
+        }
+    }
+
+    fn remove_oldest_tx(&mut self) {
+        let key = self.tree.iter().map(|(key, _)| key).next().unwrap().clone();
+        self.tree.delete(&key);
+        self.tree_size -= 1;
+    }
+
+    pub fn get(&self, id: &Nat) -> Option<TxRecord> {
+        self.tree.get(&get_key_bytes(id)).map(|v| tx_from_bytes(v))
+    }
+
+    pub fn get_range(&self, start: &Nat, limit: &Nat) -> Vec<TxRecord> {
+        fn extract_values(hash_tree: &HashTree, aggr: &mut Vec<TxRecord>) {
+            match hash_tree {
+                HashTree::Fork(boxed) => {
+                    extract_values(&boxed.as_ref().0, aggr);
+                    extract_values(&boxed.as_ref().1, aggr);
+                }
+                HashTree::Leaf(v) => {
+                    aggr.push(tx_from_bytes(&v.to_vec()));
+                }
+                HashTree::Labeled(_, child) => extract_values(child, aggr),
+                _ => {}
+            }
+        }
+
+        let witness = self.tree.value_range(
+            &get_key_bytes(&start.clone()),
+            &get_key_bytes(&(start.clone() + limit.clone() - 1)),
+        );
+        let mut result = vec![];
+        extract_values(&witness, &mut result);
+
+        result
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = TxRecord> + '_ {
+        self.tree.iter().map(|(_, value)| tx_from_bytes(value))
+    }
+
+    pub fn sign(&self) {
+        #[cfg(target_arch = "wasm32")]
+        ic_cdk::api::set_certified_data(&self.tree.root_hash());
+    }
+
+    pub fn get_witness(&self, id: &Nat) -> Option<HashTree> {
+        if self.tree.get(&get_key_bytes(id)).is_some() {
+            Some(self.tree.witness(&get_key_bytes(id)))
+        } else {
+            None
+        }
+    }
+}
+
+fn get_key_bytes(key: &Nat) -> Vec<u8> {
+    key.0.to_bytes_be()
+}
+
+fn get_tx_bytes(tx: &TxRecord) -> Vec<u8> {
+    use ic_cdk::export::candid::Encode;
+    Encode!(tx).unwrap()
+}
+
+fn tx_from_bytes(bytes: &[u8]) -> TxRecord {
+    use ic_cdk::export::candid::Decode;
+    Decode!(bytes, TxRecord).unwrap()
```

Comment:
Don't use unwrap use expect if you are sure this can't fail.
Also, can this not fail? If we ever store TxRecords somewhere and then change the struct it would certainly fail.